File: RaspberryPi_JetsonNano/python/lib/waveshare_epd/epd1in64g.py
# *****************************************************************************
# * | File        :	  epd1in64g.py
# * | Author      :   Waveshare team
# * | Function    :   Electronic paper driver
# * | Info        :
# *----------------
# * | This version:   V1
# * | Date        :   2022-07-20
# # | Info        :   python demo
# -----------------------------------------------------------------------------
# ******************************************************************************/
# Permission is hereby granted, free of charge, to any person obtaining a copy
# of this software and associated documnetation files (the "Software"), to deal
# in the Software without restriction, including without limitation the rights
# to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
# copies of the Software, and to permit persons to  whom the Software is
# furished to do so, subject to the following conditions:
#
# The above copyright notice and this permission notice shall be included in
# all copies or substantial portions of the Software.
#
# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
# IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
# FITNESS OR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
# AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
# LIABILITY WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
# OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
# THE SOFTWARE.
#

import logging
from . import epdconfig

import PIL
from PIL import Image
import io

# Display resolution
EPD_WIDTH       = 168
EPD_HEIGHT      = 168

logger = logging.getLogger(__name__)

class EPD:
    def __init__(self):
        self.reset_pin = epdconfig.RST_PIN
        self.dc_pin = epdconfig.DC_PIN
        self.busy_pin = epdconfig.BUSY_PIN
        self.cs_pin = epdconfig.CS_PIN
        self.width = EPD_WIDTH
        self.height = EPD_HEIGHT
        self.BLACK  = 0x000000   #   00  BGR
        self.WHITE  = 0xffffff   #   01
        self.YELLOW = 0x00ffff   #   10
        self.RED    = 0x0000ff   #   11
        
    # Hardware reset
    def reset(self):
        epdconfig.digital_write(self.reset_pin, 1)
        epdconfig.delay_ms(200) 
        epdconfig.digital_write(self.reset_pin, 0)         # module reset
        epdconfig.delay_ms(2)
        epdconfig.digital_write(self.reset_pin, 1)
        epdconfig.delay_ms(200)   

    def send_command(self, command):
        epdconfig.digital_write(self.dc_pin, 0)
        epdconfig.digital_write(self.cs_pin, 0)
        epdconfig.spi_writebyte([command])
        epdconfig.digital_write(self.cs_pin, 1)

    def send_data(self, data):
        epdconfig.digital_write(self.dc_pin, 1)
        epdconfig.digital_write(self.cs_pin, 0)
        epdconfig.spi_writebyte([data])
        epdconfig.digital_write(self.cs_pin, 1)
        
    def ReadBusyH(self):
        logger.debug("e-Paper busy H")
        while(epdconfig.digital_read(self.busy_pin) == 0):      # 0: idle, 1: busy
            epdconfig.delay_ms(5)
        logger.debug("e-Paper busy H release")

    def ReadBusyL(self):
        logger.debug("e-Paper busy L")
        while(epdconfig.digital_read(self.busy_pin) == 1):      # 0: busy, 1: idle
            epdconfig.delay_ms(5)
        logger.debug("e-Paper busy L release")

    def TurnOnDisplay(self):
        self.send_command(0x12) # DISPLAY_REFRESH
        self.send_data(0x01)
        self.ReadBusyH()

        self.send_command(0x02) # POWER_OFF
        self.send_data(0X00)
        self.ReadBusyH()
        
    def init(self):
        if (epdconfig.module_init() != 0):
            return -1
        # EPD hardware init start

        self.reset()

        self.send_command(0x66)
        self.send_data(0x49)
        self.send_data(0x55)
        self.send_data(0x13)
        self.send_data(0x5D)

        self.send_command(0x66)
        self.send_data(0x49)
        self.send_data(0x55)

        self.send_command(0xB0)
        self.send_data(0x03)

        self.send_command(0x00)
        self.send_data(0x4F)
        self.send_data(0x6B)

        self.send_command(0x03)
        self.send_data(0x00)

        self.send_command(0xF0)
        self.send_data(0xF6)
        self.send_data(0x0D)
        self.send_data(0x00)
        self.send_data(0x00)
        self.send_data(0x00)

        self.send_command(0x06)
        self.send_data(0xCF)
        self.send_data(0xDF)
        self.send_data(0x0F)

        self.send_command(0x41)
        self.send_data(0x00)

        self.send_command(0x50)
        self.send_data(0x30)

        self.send_command(0x60)
        self.send_data(0x0C)
        self.send_data(0x05)

        self.send_command(0x61)
        self.send_data(0xA8)
        self.send_data(0x00)
        self.send_data(0xA8)

        self.send_command(0x84)
        self.send_data(0x01)
        return 0

    def getbuffer(self, image):
        # Create a pallette with the 4 colors supported by the panel
        pal_image = Image.new("P", (1,1))
        pal_image.putpalette( (0,0,0,  255,255,255,  255,255,0,   255,0,0) + (0,0,0)*252)

        # Check if we need to rotate the image
        imwidth, imheight = image.size
        if(imwidth == self.width and imheight == self.height):
            image_temp = image
        elif(imwidth == self.height and imheight == self.width):
            image_temp = image.rotate(90, expand=True)
        else:
            logger.warning("Invalid image dimensions: %d x %d, expected %d x %d" % (imwidth, imheight, self.width, self.height))

        # Convert the soruce image to the 4 colors, dithering if needed
        image_4color = image_temp.convert("RGB").quantize(palette=pal_image)
        buf_4color = bytearray(image_4color.tobytes('raw'))

        # into a single byte to transfer to the panel
        buf = [0x00] * int(self.width * self.height / 4)
        idx = 0
        for i in range(0, len(buf_4color), 4):
            buf[idx] = (buf_4color[i] << 6) + (buf_4color[i+1] << 4) + (buf_4color[i+2] << 2) + buf_4color[i+3]
            idx += 1

        return buf

    def display(self, image):
        if self.width % 4 == 0 :
            Width = self.width // 4
        else :
            Width = self.width // 4 + 1
        Height = self.height

        self.send_command(0x68)
        self.send_data(0x01)

        self.send_command(0x04)
        self.ReadBusyH()

        self.send_command(0x10)
        for j in range(0, Height):
            for i in range(0, Width):
                    self.send_data(image[i + j * Width])

        self.send_command(0x68)
        self.send_data(0x00)

        self.TurnOnDisplay()
        
    def Clear(self, color=0x55):
        if self.width % 4 == 0 :
            Width = self.width // 4
        else :
            Width = self.width // 4 + 1
        Height = self.height

        self.send_command(0x68)
        self.send_data(0x01)

        self.send_command(0x04)
        self.ReadBusyH()

        self.send_command(0x10)
        for j in range(0, Height):
            for i in range(0, Width):
                self.send_data(color)

        self.send_command(0x68)
        self.send_data(0x00)

        self.TurnOnDisplay()

    def sleep(self):
        self.send_command(0x02) # POWER_OFF
        self.send_data(0x00)

        self.send_command(0x07) # DEEP_SLEEP
        self.send_data(0XA5)
        
        epdconfig.delay_ms(2000)
        epdconfig.module_exit()
### END OF FILE ###

File: STM32/STM32-F103ZET6/User/e-Paper/EPD_1in64g.c
/*****************************************************************************
* | File      	:   EPD_1in64_g.c
* | Author      :   Waveshare team
* | Function    :   1.64inchg e-paper (G)
* | Info        :
*----------------
* |	This version:   V1.0
* | Date        :   2022-07-22
* | Info        :
* -----------------------------------------------------------------------------
#
# Permission is hereby granted, free of charge, to any person obtaining a copy
# of this software and associated documnetation files (the "Software"), to deal
# in the Software without restriction, including without limitation the rights
# to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
# copies of the Software, and to permit persons to  whom the Software is
# furished to do so, subject to the following conditions:
#
# The above copyright notice and this permission notice shall be included in
# all copies or substantial portions of the Software.
#
# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
# IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
# FITNESS OR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
# AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
# LIABILITY WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
# OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
# THE SOFTWARE.
#
******************************************************************************/
#include "EPD_1in64g.h"
#include "Debug.h"

/******************************************************************************
function :	Software reset
parameter:
******************************************************************************/
static void EPD_1IN64G_Reset(void)
{
    DEV_Digital_Write(EPD_RST_PIN, 1);
    DEV_Delay_ms(20);
    DEV_Digital_Write(EPD_RST_PIN, 0);
    DEV_Delay_ms(1);
    DEV_Digital_Write(EPD_RST_PIN, 1);
    DEV_Delay_ms(20);
}

/******************************************************************************
function :	send command
parameter:
     Reg : Command register
******************************************************************************/
static void EPD_1IN64G_SendCommand(UBYTE Reg)
{
    DEV_Digital_Write(EPD_DC_PIN, 0);
    DEV_Digital_Write(EPD_CS_PIN, 0);
    DEV_SPI_WriteByte(Reg);
    DEV_Digital_Write(EPD_CS_PIN, 1);
}

/******************************************************************************
function :	send data
parameter:
    Data : Write data
******************************************************************************/
static void EPD_1IN64G_SendData(UBYTE Data)
{
    DEV_Digital_Write(EPD_DC_PIN, 1);
    DEV_Digital_Write(EPD_CS_PIN, 0);
    DEV_SPI_WriteByte(Data);
    DEV_Digital_Write(EPD_CS_PIN, 1);
}

/******************************************************************************
function :	Wait until the busy_pin goes LOW
parameter:
******************************************************************************/
static void EPD_1IN64G_ReadBusyH(void)
{
    Debug("e-Paper busy H\r\n");
    while(!DEV_Digital_Read(EPD_BUSY_PIN)) {      //LOW: idle, HIGH: busy
        DEV_Delay_ms(5);
    }
    Debug("e-Paper busy H release\r\n");
}

/******************************************************************************
function :	Turn On Display
parameter:
******************************************************************************/
static void EPD_1IN64G_TurnOnDisplay(void)
{
    EPD_1IN64G_SendCommand(0x12); // DISPLAY_REFRESH
    EPD_1IN64G_SendData(0x01);
    EPD_1IN64G_ReadBusyH();

    EPD_1IN64G_SendCommand(0x02); // POWER_OFF
    EPD_1IN64G_SendData(0X00);
    EPD_1IN64G_ReadBusyH();
}

/******************************************************************************
function :	Initialize the e-Paper register
parameter:
******************************************************************************/
void EPD_1IN64G_Init(void)
{
    EPD_1IN64G_Reset();

    EPD_1IN64G_SendCommand(0x66);
    EPD_1IN64G_SendData(0x49);
    EPD_1IN64G_SendData(0x55);
    EPD_1IN64G_SendData(0x13);
    EPD_1IN64G_SendData(0x5D);

    EPD_1IN64G_SendCommand(0x66);
    EPD_1IN64G_SendData(0x49);
    EPD_1IN64G_SendData(0x55);

    EPD_1IN64G_SendCommand(0xB0);
    EPD_1IN64G_SendData(0x03);//1 boost 20211113


    EPD_1IN64G_SendCommand(0x00);
    EPD_1IN64G_SendData(0x4F);
    EPD_1IN64G_SendData(0x6B);

    EPD_1IN64G_SendCommand(0x03);
    EPD_1IN64G_SendData(0x00);

    EPD_1IN64G_SendCommand(0xF0);
    EPD_1IN64G_SendData(0xF6);
    EPD_1IN64G_SendData(0x0D);
    EPD_1IN64G_SendData(0x00);
    EPD_1IN64G_SendData(0x00);
    EPD_1IN64G_SendData(0x00);

    //20220303
    EPD_1IN64G_SendCommand(0x06);
    EPD_1IN64G_SendData(0xCF);
    EPD_1IN64G_SendData(0xDF);
    EPD_1IN64G_SendData(0x0F);

    EPD_1IN64G_SendCommand(0x41);
    EPD_1IN64G_SendData(0x00);

    EPD_1IN64G_SendCommand(0x50);
    EPD_1IN64G_SendData(0x30);

    EPD_1IN64G_SendCommand(0x60);
    EPD_1IN64G_SendData(0x0C); 
    EPD_1IN64G_SendData(0x05);

    EPD_1IN64G_SendCommand(0x61);
    EPD_1IN64G_SendData(0xA8);
    EPD_1IN64G_SendData(0x00); 
    EPD_1IN64G_SendData(0xA8); 

    EPD_1IN64G_SendCommand(0x84);
    EPD_1IN64G_SendData(0x01);

}

/******************************************************************************
function :	Clear screen
parameter:
******************************************************************************/
void EPD_1IN64G_Clear(UBYTE color)
{
    UWORD Width, Height;
    Width = (EPD_1IN64G_WIDTH % 4 == 0)? (EPD_1IN64G_WIDTH / 4 ): (EPD_1IN64G_WIDTH / 4 + 1);
    Height = EPD_1IN64G_HEIGHT;

    EPD_1IN64G_SendCommand(0x68);
    EPD_1IN64G_SendData(0x01);
    
    EPD_1IN64G_SendCommand(0x04);
    EPD_1IN64G_ReadBusyH();

    EPD_1IN64G_SendCommand(0x10);
    for (UWORD j = 0; j < Height; j++) {
        for (UWORD i = 0; i < Width; i++) {
            EPD_1IN64G_SendData((color << 6) | (color << 4) | (color << 2) | color);
        }
    }

    EPD_1IN64G_SendCommand(0x68);
    EPD_1IN64G_SendData(0x00);

    EPD_1IN64G_TurnOnDisplay();
}

/******************************************************************************
function :	Sends the image buffer in RAM to e-Paper and displays
parameter:
******************************************************************************/
void EPD_1IN64G_Display(UBYTE *Image)
{
    UWORD Width, Height;
    Width = (EPD_1IN64G_WIDTH % 4 == 0)? (EPD_1IN64G_WIDTH / 4 ): (EPD_1IN64G_WIDTH / 4 + 1);
    Height = EPD_1IN64G_HEIGHT;

    EPD_1IN64G_SendCommand(0x68);
    EPD_1IN64G_SendData(0x01);
    
    EPD_1IN64G_SendCommand(0x04);
    EPD_1IN64G_ReadBusyH();

    EPD_1IN64G_SendCommand(0x10);
    for (UWORD j = 0; j < Height; j++) {
        for (UWORD i = 0; i < Width; i++) {
            EPD_1IN64G_SendData(Image[i + j * Width]);
        }
    }

    EPD_1IN64G_SendCommand(0x68);
    EPD_1IN64G_SendData(0x00);

    EPD_1IN64G_TurnOnDisplay();
}

/******************************************************************************
function :	Enter sleep mode
parameter:
******************************************************************************/
void EPD_1IN64G_Sleep(void)
{
    EPD_1IN64G_SendCommand(0x02); // POWER_OFF
    EPD_1IN64G_SendData(0X00);
    EPD_1IN64G_SendCommand(0x07); // DEEP_SLEEP
    EPD_1IN64G_SendData(0XA5);
}

